tool/abi.rb

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env ruby
+
+class ABI
+  def initialize(srcdir)
+    @srcdir = srcdir
+    @output_file = "#{@srcdir}/include/ruby/internal/abi.h"
+  end
+
+  attr_reader :output_file
+
+  def version
+    # Generate MD5 for every header file
+    header_files = Dir["#{@srcdir}/include/**/*.h"].reject { |f| f == output_file }
+    md5s = header_files.map do |file|
+      md5_hexdigest(File.read(file))
+    end
+
+    # Generate MD5 of MD5s
+    md5 = md5_hexdigest(md5s.join)
+
+    # Take only bottom 8 bytes of 16 byte MD5 string
+    md5.to_i(16) % (2**64)
+  end
+
+  private
+
+  def md5_hexdigest(str)
+    begin
+      require "digest"
+    rescue LoadError
+      return "0"
+    end
+
+    Digest::MD5.hexdigest(str)
+  end
+end
+
+# Run only if file was directly called
+if __FILE__ == $0
+  require "optparse"
+  require_relative "lib/colorize"
+
+  OptionParser.new do |opts|
+    opts.on("--srcdir=PATH", "use PATH as source directory") do |srcdir|
+      @srcdir = srcdir
+    end
+  end.parse!
+
+  abi = ABI.new(@srcdir)
+  color = Colorize.new(color)
+  updated = color.fail("updated")
+
+  abi_h = <<-C
+  #ifndef RUBY_ABI_H
+  #define RUBY_ABI_H
+
+  // DO NOT MODIFY THIS FILE DIRECTLY
+  // This file was automatically generated by tool/abi.rb
+
+  unsigned long long
+  #ifndef RUBY_DLN_CHECK_ABI
+  __attribute__((weak))
+  #endif
```

Comment:
Does this mean it's weak only if we are in a release-or-windows?
Shouldn't it just always be weak, except on Windows where it's not supported?

Reply:
That's a good point. I think checking `RUBY_DLN_CHECK_ABI` here can be completely removed since I think Windows ignores the `__attribute__((weak))` completely.